### PR TITLE
Langjian/distributed json 

### PR DESCRIPTION
**Pull request by jianyinglang**

### CODE REVIEW

File: examples/mnist/mnist_softmax_distributed.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  #
+  # So here we use tf.nn.softmax_cross_entropy_with_logits on the raw
+  # outputs of 'y', and then average across the batch.
+  cross_entropy = tf.reduce_mean(
+      tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y))
+  #global_step = tf.train.get_or_create_global_step()
+  global_step = tf.contrib.framework.get_or_create_global_step()
+  opt = tf.train.GradientDescentOptimizer(0.5)
+  # Add MPI Distributed Optimizer
+  #import pdb; pdb.set_trace()
```
Comment (sayantan-nervana) [inline examples/mnist/mnist_softmax_distributed.py]: this `pub` line can be removed

File: examples/mnist/mnist_softmax_distributed.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    print("Training time: %f seconds" % (end-start))
+
+if __name__ == '__main__':
+  parser = argparse.ArgumentParser()
+  parser.add_argument('--data_dir', type=str, default='/tmp/tensorflow/mnist/input_data',
+                      help='Directory for storing input data')
+  parser.add_argument('--log_dir',type=str, default='/tmp/tensorflow/mnist/logs/mnist_with_summaries',
+                      help='Summaries log directory')
+  FLAGS, unparsed = parser.parse_known_args()
+  tf.app.run(main=main, argv=[sys.argv[0]] + unparsed)
```
Comment (sayantan-nervana) [inline examples/mnist/mnist_softmax_distributed.py]: Can you please add the command line to run this as a comment in the end of the file

File: src/ngraph_encapsulate_op.cc [hunk 3]
```diff
@@ -39,6 +39,10 @@
 
 #include "ngraph/runtime/interpreter/int_backend.hpp"
 
+#ifdef NGRAPH_DISTRIBUTED
+#include <mpi.h>>
```
Comment (sayantan-nervana) [inline src/ngraph_encapsulate_op.cc]: `>>`. Is this a typo?

File: src/ngraph_rewrite_pass.cc [hunk 4]
```diff
@@ -29,6 +29,10 @@
 
 #include <iomanip>
 
+#ifdef NGRAPH_DISTRIBUTED
+#include <mpi.h>>
```
Comment (sayantan-nervana) [inline src/ngraph_rewrite_pass.cc]: `>>` is this a typo?

Comment (jianyinglang) [reply src/ngraph_rewrite_pass.cc]: Yes

File: src/ngraph_encapsulate_op.cc (hunk 3 above)
Comment (jianyinglang) [reply src/ngraph_encapsulate_op.cc]: Yes

File: src/ngraph_rewrite_pass.cc (hunk 4 above)
Comment (jianyinglang) [reply src/ngraph_rewrite_pass.cc]: Thanks a lot for catching this!

File: examples/mnist/mnist_softmax_distributed.py (hunk 2 above)
Comment (jianyinglang) [reply examples/mnist/mnist_softmax_distributed.py]: Added an example run command for distributed runs

File: examples/mnist/mnist_softmax_distributed.py (hunk 1 above)
Comment (jianyinglang) [reply examples/mnist/mnist_softmax_distributed.py]: deleted

File: examples/mnist/mnist_softmax_distributed.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        default='/tmp/tensorflow/mnist/input_data',
+        help='Directory for storing input data')
+    parser.add_argument(
+        '--log_dir',
+        type=str,
+        default='/tmp/tensorflow/mnist/logs/mnist_with_summaries',
+        help='Summaries log directory')
+    FLAGS, unparsed = parser.parse_known_args()
+    tf.app.run(main=main, argv=[sys.argv[0]] + unparsed)
+# run comand for this distributed script
```
[the remark this answers is not in the document]
Comment (jianyinglang) [reply examples/mnist/mnist_softmax_distributed.py]: Thanks for catching this. Fixed it.

File: build_ngtf.py [hunk 6]
```diff
@@ -512,6 +512,7 @@ def main():
         "-DNGRAPH_TUNE_ARCH=" + target_arch,
         "-DNGRAPH_ARTIFACTS_DIR=" + artifacts_location,
         "-DUNIT_TEST_ENABLE=ON",
+        "-DNGRAPH_DISTRIBUTED_ENABLE=TRUE",
```
Comment (avijit-nervana) [inline build_ngtf.py]: The default should be false i.e., for regular builds no need to enable distributed. Need to add a command line option to the build_ngtf.py (e.g., `--enable_distributed`) which when specified will add this flag. See the --debug_build as an example of how to extend the cmake options.

File: examples/mnist/mnist_softmax_distributed.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""A very simple MNIST classifier.
+
+See extensive documentation at
+https://www.tensorflow.org/get_started/mnist/beginners
+"""
+from __future__ import absolute_import
```
Comment (avijit-nervana) [inline examples/mnist/mnist_softmax_distributed.py]: Please provide a summary of modifications made to this file and a reference to the source of this file.

File: src/ngraph_encapsulate_op.cc [hunk 8]
```diff
@@ -36,6 +36,10 @@
 
 #include "ngraph/runtime/interpreter/int_backend.hpp"
 
+#ifdef NGRAPH_DISTRIBUTED
```
Comment (avijit-nervana) [inline src/ngraph_encapsulate_op.cc]: It's better to use the `if defined(NGRAPH_DISTRIBUTED)`

File: src/ngraph_rewrite_pass.cc [hunk 9]
```diff
@@ -102,13 +106,33 @@ class NGraphRewritePass : public GraphOptimizationPass {
   static std::string GraphFilenamePrefix(std::string kind, int idx) {
     std::stringstream ss;
     ss << kind << "_" << std::setfill('0') << std::setw(4) << idx;
+#ifdef NGRAPH_DISTRIBUTED
```
Comment (avijit-nervana) [inline src/ngraph_rewrite_pass.cc]: Consider a function that can be used instead of the code repetition below.

File: CMakeLists.txt [hunk 10]
```diff
@@ -226,12 +226,12 @@ message(STATUS "NGRAPH_PLAIDML_ENABLE:      ${NGRAPH_PLAIDML_ENABLE}")
 message(STATUS "NGRAPH_TARGET_ARCH:         ${NGRAPH_TARGET_ARCH}")
 message(STATUS "NGRAPH_TUNE_ARCH:           ${NGRAPH_TUNE_ARCH}")
 
-if(NGRAPH_DISTRIBUTED_ENABLE)
-    find_package(MPI REQUIRED)
-    add_definitions(-DNGRAPH_DISTRIBUTED)
-    include_directories(SYSTEM ${MPI_C_INCLUDE_PATH} ${MPI_CXX_INCLUDE_PATH})
-    link_directories(${MPI_C_LIBRARIES} ${MPI_CXX_LIBRARIES})
-endif()
+#if(NGRAPH_DISTRIBUTED_ENABLE)
```
Comment (avijit-nervana) [inline CMakeLists.txt]: Is this temporary? If not please remove the commented out code.

File: build_ngtf.py (hunk 6 above)
Comment (jianyinglang) [reply build_ngtf.py]: Added the option of `--distributed_build` for distributed build

File: examples/mnist/mnist_softmax_distributed.py (hunk 7 above)
Comment (jianyinglang) [reply examples/mnist/mnist_softmax_distributed.py]: Listed the changes I made from the source file and added the link to the reference source of this file

File: src/ngraph_encapsulate_op.cc (hunk 8 above)
Comment (jianyinglang) [reply src/ngraph_encapsulate_op.cc]: Changed to `#if defined NGRAPH_DISTRIBUTED`

File: src/ngraph_rewrite_pass.cc (hunk 9 above)
Comment (jianyinglang) [reply src/ngraph_rewrite_pass.cc]: Used the distributed class defined in nGraph core

**Comment by jianyinglang** [comment]
TESTNOW

**Pull request merged**